### DISCUSSION TITLE
What a proper way of upgrading HA cluster (Round-robin DNS record as Fixed Registration Address and Kubernetes API entrypoint)

### DISCUSSION THREAD

**Discussion by agrevtsev** [rancher/rke2#6509] [Q&A] [answered]
Hi team!
We have couple of HA clusters, which have round-robin DNS record as Fixed Registration Address and Kubernetes API entrypoint.
We planing to upgrade those, and there's come question which bothers me much:
As per [k8s version skew policy](https://kubernetes.io/releases/version-skew-policy/#kube-controller-manager-kube-scheduler-and-cloud-controller-manager), neither kube-controller-manager, kube-scheduler, and cloud-controller-manager should speak with an older kube-apiserver version.
But as i understand, all of mentioned CP components are upgraded simultaneously, as `systemctl restart rke2-server.service` is run.
So could you suggest some upgrade path, which would allow to:
- evade downtime during upgrade
- prevent breaking version skew policy (iow preventing newer version CP components connecting to an older version kube-apiserver)
Or maybe i'm overthink?

Thanks for your time!

**Comment by brandond** [comment, marked as answer]
Just upgrade the servers one at a time, following the version skew policy docs. You can do this manually or with a system upgrade controller job. The Kubernetes components on the control-plane node all talk to their local apiserver, which will be upgraded when the server is restarted.
